### ISSUE TITLE
Highlighted code should be wrapped in code element

### ISSUE THREAD

**Issue by marcoscaceres** [w3c/respec#2098]
Code coming back from the syntax highlighter should be wrapped in a `<code>` element, so to semantically conform with HTML spec.

**Comment by marcoscaceres** [comment]
Needs to happen around here:
https://github.com/w3c/respec/blob/develop/src/core/highlight.js#L61

Could be as simple as:
```JS
element.innerHTML = `<code>${value}</code>`;
```

**Comment by marcoscaceres** [comment]
Note, the above is only for `pre` elements... so actually, it would need to wrap.

**Comment by saschanaz** [comment]
Maybe we should also allow the standard-styled input: `<pre><code style="(lang)">... code ...</code></pre>`.

**Comment by marcoscaceres** [comment]
Yeah, I would not mind that at all.

**Comment by CodHeK** [comment]
@marcoscaceres that issue got closed, can I take this one?

**Comment by marcoscaceres** [comment]
@CodHeK, sure. But we are still working out the details. We know at a minimum we want to wrap the pre in code, so that might be a place to start.

**Comment by marcoscaceres** [comment]
Oops, I mean wrap the highlighted code in in a `code`, not "the pre in code".

**Comment by CodHeK** [comment]
@marcoscaceres I don't follow, why should the contents in `<pre` be wrapped using the `<code>` tag ?

**Comment by marcoscaceres** [comment]
Please see https://github.com/w3c/respec/issues/1970#issuecomment-464538998

**Comment by saschanaz** [comment]
From https://github.com/w3c/respec/issues/1970#issuecomment-465433018:

>Does this mean, if the `<pre>` element has `class="nohighlight"` we need not wrap the contents of the `pre` inside `code` ?

I'd say yes.

**Comment by CodHeK** [comment]
Does this mean, if the `<pre>` element has `class="nohighlight"` we need not wrap the contents of the `pre` inside `code` ? And wrap it other case when some other class or no class is used for the pre tag.

**Comment by CodHeK** [comment]
@saschanaz but, then in cases such as these ...

```
<pre>
     <a>Foo</a>
      const { op } = this.state;
      for(let i = 0; i < t.length; i++) {
          console.log(i);
       }
</pre>
```

**Comment by CodHeK** [comment]
We'd have to check if it the `pre` is a pure `code` element or not and warp it accordingly

**Comment by saschanaz** [comment]
I doubt anyone would mix hyperlinks and actual codes within a single element... Have you seen such cases? @marcoscaceres

**Comment by marcoscaceres** [comment]
What @marcoscaceres said. I don't think we should support linking in those case. If we do eventually support linking, we should do that in as part of a different issue.

**Comment by CodHeK** [comment]
So, I guess the only change here would to be add an if condition right ?

```
if(element.classList.length > 0)
          element.innerHTML = value;
 else
          element.innerHTML = `<code>${value}</code>`;
```

**Comment by saschanaz** [comment]
I hope we do slightly more: https://github.com/w3c/respec/issues/2098#issuecomment-464547954

In this case, we check whether a `<pre>` have a  `<code>` as its sole child. If it does we should use `code` inside `pre` to highlight, otherwise we should use `<pre>`.

**Comment by saschanaz** [comment]
With `<code>` Maybe we could do:

```
<pre>
  <a>Foo</a> // We have <code> here so do not pass this to the highlighter
  <code class="js">class Foo {}</code> // Highlighted
</pre>

**Comment by marcoscaceres** [comment]
Yeah, that makes sense. Might also have multiple code blocks?

**Comment by CodHeK** [comment]
> With `<code>` maybe we could do:
> 
> ```
> <pre>
>   <a>Foo</a> // We have `code` here so do not pass this to the highlighter
>   <code class="js">class Foo {}</code> // Highlighted
> </pre>
> ```

@saschanaz  Okay, so, what we need to do here is : 
 
Check if `<pre>` has `<code>` in it, if yes then it gets highlighted according to the class in `<code>` else we wrap the content of `<pre>` in `<code>` right ?

**Comment by saschanaz** [comment]
>Basically, we'd have to check for `<code>` elements inside `<pre>` if none exist we just wrap all of the content of `<pre>`.

@CodHeK Exactly. Would be simpler if we do the wrapping first and then highlight for each `<code>` inside.

**Comment by CodHeK** [comment]
@saschanaz is this the proper highlighting for the case you mentioned above ...

![screenshot 2019-02-20 at 5 41 24 pm](https://user-images.githubusercontent.com/22097366/53090993-df28a600-3536-11e9-8e1e-80a3fcd72068.png)

**Comment by saschanaz** [comment]
`//` parts shouldn't be highlighted as comments as they are not in `<code>`.

**Comment by CodHeK** [comment]
Ah, sorry it's for this case, 

```
<pre>
  <a>Foo</a> // We have `code` here so do not pass this to the highlighter
  class Foo {}// Highlighted
</pre>
```

It's when, the content gets wrapped in `<code>`

**Comment by saschanaz** [comment]
Ah yes, looks good then.

**Comment by CodHeK** [comment]
@saschanaz sent a PR that'll just wrap if node `<code>` is found within `<pre>` #2111

**Comment by marcoscaceres** [comment]
Fixed by #2132
